### ISSUE TITLE
New expiration strategy

### ISSUE THREAD

**Issue by joekerna** [laurent22/rsync-time-backup#105]
This concerns the branch expiration-strategy:
I create a backup every day at noon. I've set the strategy parameter to `--strategy "365:7 150:3 1:1"`
Today (Dec 14th) I have backups from the following days:
```
drwxr-xr-x 7 root root 4,0K Apr 25  2017 2017-04-30-222401
drwxr-xr-x 7 root root 4,0K Apr 25  2017 2017-05-10-220002
drwxr-xr-x 7 root root 4,0K Jun 29 19:08 2017-06-30-120001
drwxr-xr-x 7 root root 4,0K Jul 18 17:47 2017-07-31-120001
drwxr-xr-x 7 root root 4,0K Aug 29 11:08 2017-08-31-120001
drwxr-xr-x 7 root root 4,0K Sep 26 19:44 2017-09-30-120001
drwxr-xr-x 7 root root 4,0K Okt 19 20:03 2017-11-15-120001
drwxr-xr-x 7 root root 4,0K Okt 19 20:03 2017-11-16-120001
drwxr-xr-x 7 root root 4,0K Okt 19 20:03 2017-11-17-120001
drwxr-xr-x 7 root root 4,0K Okt 19 20:03 2017-11-18-120001
drwxr-xr-x 7 root root 4,0K Okt 19 20:03 2017-11-19-120001
drwxr-xr-x 7 root root 4,0K Okt 19 20:03 2017-11-20-120001
drwxr-xr-x 7 root root 4,0K Okt 19 20:03 2017-11-21-120001
drwxr-xr-x 7 root root 4,0K Nov 22 10:33 2017-11-22-120001
drwxr-xr-x 7 root root 4,0K Nov 22 10:33 2017-11-23-120002
drwxr-xr-x 7 root root 4,0K Nov 22 10:33 2017-11-24-201549
drwxr-xr-x 7 root root 4,0K Nov 22 10:33 2017-11-25-120001
drwxr-xr-x 7 root root 4,0K Nov 22 10:33 2017-11-26-120002
drwxr-xr-x 7 root root 4,0K Nov 22 10:33 2017-11-27-191832
drwxr-xr-x 7 root root 4,0K Nov 22 10:33 2017-11-28-120001
drwxr-xr-x 7 root root 4,0K Nov 22 10:33 2017-11-29-120001
drwxr-xr-x 7 root root 4,0K Nov 22 10:33 2017-11-30-120001
drwxr-xr-x 7 root root 4,0K Nov 22 10:33 2017-12-01-120002
drwxr-xr-x 7 root root 4,0K Nov 22 10:33 2017-12-02-120001
drwxr-xr-x 7 root root 4,0K Nov 22 10:33 2017-12-03-120001
drwxr-xr-x 7 root root 4,0K Nov 22 10:33 2017-12-04-120001
drwxr-xr-x 7 root root 4,0K Nov 22 10:33 2017-12-05-120001
drwxr-xr-x 7 root root 4,0K Nov 22 10:33 2017-12-06-120001
drwxr-xr-x 7 root root 4,0K Nov 22 10:33 2017-12-07-120001
drwxr-xr-x 7 root root 4,0K Dez  8 07:46 2017-12-08-120001
drwxr-xr-x 7 root root 4,0K Dez  8 07:46 2017-12-09-220124
drwxr-xr-x 7 root root 4,0K Dez  8 07:46 2017-12-10-132448
drwxr-xr-x 7 root root 4,0K Dez  8 07:46 2017-12-11-074832
drwxr-xr-x 7 root root 4,0K Dez  8 07:46 2017-12-12-074832
drwxr-xr-x 7 root root 4,0K Dez  8 07:46 2017-12-13-083815
drwxr-xr-x 7 root root 4,0K Dez  8 07:46 2017-12-14-074832
```
I'm missing at least one backup from October. I know that at some point I've had a backup from October 31st.
From my understanding `150:3` means that for the last 150, one backup is kept every three days.
At one point the backup from October 31st was deleted though.
Do you have a script that can emulate that history?

**Comment by joekerna** [comment]
This is a significant bug! I kept losing more and more backups and have reverted to the old version for now.

**Comment by laurent22** [comment]
Hi, sorry for the late reply, I've finally had time to look into this.

It looks like the strategy flag was not implemented at all so it's always using the default strategy (which, based on your test above, is working fine). I'm certain i had implemented this flag though but maybe i forgot to commit the changes. I'll take a look and fix this.

The good news is that the strategy algorithm is working fine. By default it's:
- after 365 days, keep one backup every 30 days (365:30)
- After 30 days keep one backup every 7 days (30:7)
- After 1 day keep one backup every 1 day (1:1)

And that correspond to the tests you've done.

**Comment by laurent22** [comment]
Yes I had unfortunately forgotten to push the latest changes, so the branch was incomplete and in particular did not handle the `--strategy` flag.

The latest changes are now on [the branch](https://github.com/laurent22/rsync-time-backup/tree/expiration-strategy) along with a script to test the strategy.

For example, here are two tests with backup sets created every 12h and processed through the expiration logic:

**1:1 30:7 365:30**

	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2016-11-01-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2016-12-01-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2016-12-31-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-01-30-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-03-01-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-03-08-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-03-15-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-03-22-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-03-30-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-04-06-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-04-13-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-04-20-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-04-27-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-05-04-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-05-11-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-05-18-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-05-25-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-06-01-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-06-08-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-06-15-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-06-22-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-06-29-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-07-06-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-07-13-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-07-20-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-07-27-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-08-03-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-08-10-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-08-17-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-08-24-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-08-31-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-09-07-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-09-14-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-09-21-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-09-28-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-10-05-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-10-12-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-10-19-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-10-26-161819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-11-02-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-11-09-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-11-16-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-11-23-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-11-30-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-12-07-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-12-14-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-12-21-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2017-12-28-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-01-04-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-01-11-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-01-18-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-01-25-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-01-26-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-01-27-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-01-28-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-01-29-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-01-30-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-01-31-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-01-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-02-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-03-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-04-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-05-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-06-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-07-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-08-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-09-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-10-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-11-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-12-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-13-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-14-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-15-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-16-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-17-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-18-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-19-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-20-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-21-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-22-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-23-031819
	drwxrwxrwx 0 root root 4.0K Feb 23 17:18 2018-02-23-151819
	drwxrwxrwx 0 root root 4.0K Feb 23 16:38 2018-02-23-171819

**1:1 7:2 30:10**

	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2016-10-12-171608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2016-10-22-171608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2016-11-01-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2016-11-11-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2016-11-21-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2016-12-01-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2016-12-11-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2016-12-21-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2016-12-31-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-01-10-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-01-20-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-01-30-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-02-09-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-02-19-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-03-01-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-03-11-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-03-21-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-04-01-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-04-11-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-04-21-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-05-01-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-05-11-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-05-21-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-05-31-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-06-10-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-06-20-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-06-30-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-07-10-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-07-20-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-07-30-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-08-09-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-08-19-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-08-29-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-09-08-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-09-18-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-09-28-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-10-08-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-10-18-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-10-28-161608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-11-07-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-11-17-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-11-27-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-12-07-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-12-17-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2017-12-27-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-01-06-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-01-16-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-01-26-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-01-28-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-01-30-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-02-01-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-02-03-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-02-05-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-02-07-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-02-09-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-02-11-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-02-13-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-02-15-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-02-17-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-02-18-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-02-19-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-02-20-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-02-21-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-02-22-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-02-23-031608
	drwxrwxrwx 0 root root 4.0K Feb 23 17:16 2018-02-23-151608
	drwxrwxrwx 0 root root 4.0K Feb 23 16:38 2018-02-23-171608
	
So it seems to be working fine. I will start using it for my backups and merge it to master in a few weeks if everything works fine.

**Comment by neddy70** [comment]
Hi Guys, 

The new expiration strategy looks good, I like the flexibility.

I'm getting an error when parsing the dates:
```
rsync_tmbackup: [WARNING] Could not parse date: /Users/ned/Desktop/EFI_Bak2//2018-03-18-124508
Failed conversion of ``2018-03-18'' using format ``%Y-%m-%d-%H%M%S''
date: illegal time format
usage: date [-jnRu] [-d dst] [-r seconds] [-t west] [-v[+|-]val[ymwdHMS]] ...
            [-f fmt date | [[[mm]dd]HH]MM[[cc]yy][.ss]] [+format]
```

I'm on macOS 10.12, it looks like on [line 71](https://github.com/laurent22/rsync-time-backup/blob/4af7610dd15a0d8a9e2755f202fd29dc312cf8eb/rsync_tmbackup.sh#L71) the `date_format` passed to `fn_parse_date()` is not being used. I would have liked to fix it, but I thought it was a bit beyond my skill level, and I assume there was a plan to implement this, so I thought I would leave it. 


On another note:

I had a need to add the option to turn off the automatic purging of old backups when out of space, so I create a fork with a branch to implemented it: [my branch](https://github.com/neddy70/rsync-time-backup/tree/no-auto-expire). 

It does not change the default behaviour of the script, it just allows an extra option to be passed to the script to turn it off. I have had users in the past mistakenly add a large amount of data to the backup source, and then this is turn wipes out all backup history, when in reality I would rather have just had the script error, and then I could fix it.

Would you be interested in a pull request to add this?

**Comment by laurent22** [comment]
@neddy70, please pull the latest changes, I had forgotten to push one commit (which had a fix for the parse_date function). Yes I think your change could be useful, feel free to create a pull request for it.

**Comment by joekerna** [comment]
I can confirm the correct behavior of the branch.

Thanks a lot!

**Comment by yorkday** [comment]
I like the features in this branch, however I see a couple of issues.

1. Previous strategy in main branch was to keep most recent backup each day for last 31 days.  When I used the new default expiration strategy, it removed backups that were not exactly 1 day older than the previous backup.  Due to the differences when the job runs by the scheduler, there is small variation in the timestamps.  The result is backups that are seconds less than 1 day old are removed, so there is no backup kept for that day.

Example master branch:
2018-04-23-000007
2018-04-24-000006
2018-04-25-000007
2018-04-26-000027 <-- This backup becomes removed by new branch
2018-04-27-000007
2018-04-28-000007
2018-04-29-000007

Example this branch:
2018-04-23-000006
2018-04-24-000006
2018-04-25-000007
2018-04-27-000007
2018-04-28-000007
2018-04-29-000007

2. The documentation does not make it clear where the --strategy or --no-auto-expire flags should be placed.  Originally I placed them in the string for --rsync-set-flags, however I got an error and realised that rsync itself doesn't expect these two flags used by rsync-time-backup.  Clarity in the readme where these additional options should be placed would be appreciated.
